Close #1 Create Teleop Node

matthewia94 opened the pull request:
I have created a very simplistic version of a teleop node that should allow us to control the robot in an arcade drive.  It is currently set up to work only with wireless xbox 360 controllers.

matthewia94 committed:
Add control package

The control package will contain nodes which relate to robot control.
This will include nodes such as the TeleOp node, the twist_mux node and
the motor controller node.

matthewia94 committed:
Add sensor_msgs package dependency

The joy message is contained in the sensor_msgs package and so
sensor_msgs should be a build and run dependancy.  It has been added to
both the package manifest and the CMakeList

matthewia94 committed:
Add remote_control_node

Add a remote control_node which simply converts joy messages on the
topic "joy" to a twist message and publishes it on the topic
"teleop_twist".

Closes #1

matthewia94 commented on ros_ws/src/control/src/remote_control.cpp:
```diff
@@ -0,0 +1,49 @@
+void RemoteControl::joyCallback(const sensor_msgs::Joy::ConstPtr& msg)
+{
+  cmd_vel.linear.x = msg->axes[1];
+  cmd_vel.angular.z = msg->axes[2];
```
Any suggestions on how to make this a non magic number?  One way would be defines in the header which is what we have done previously.  This can change based on the controller we are using so maybe we should set up some way to query what controller uses what button/axes map.

islamelnabarawy replied:
This is basically the input mapping for the controller, right? It can change based on the controller and we may even need to change it for the same controller (e.g. to improve usability).

Using constants or even literals can work for now, but we need to design a flexible controller input mapping mechanism and create a new task for that. A parameter file is probably a good fit for this mechanism.

matthewia94 committed:
Fix small tabbing issue

The default constructor was using two tabs instead of one.  Should be
uniform throughout the file now.

islamelnabarawy commented on ros_ws/src/control/CMakeLists.txt:
```diff
@@ -0,0 +1,191 @@
+##   * uncomment the add_*_files sections below as needed
+##     and list every .msg/.srv/.action file to be processed
+##   * uncomment the generate_messages entry below
+##   * add every package in MSG_DEP_SET to generate_messages(DEPENDENCIES ...)
```
It's probably a good idea to trim all these instructions and commented out configurations. It will make the file a lot easier to edit in the future, and we can always look up those instructions if we needed to make changes. They don't need to be repeated in every CMakeLists file we have.

We can keep the section divider comments, and anything else that is unique to this file.

islamelnabarawy commented on ros_ws/src/control/package.xml:
```diff
@@ -0,0 +1,60 @@
+  <!-- One maintainer tag required, multiple allowed, one person per tag --> 
+  <!-- Example:  -->
+  <!-- <maintainer email="jane.doe@example.com">Jane Doe</maintainer> -->
+  <maintainer email="matt@todo.todo">matt</maintainer>
```
Please fix this to use a valid email address, and remove all the unnecessary comment tags (see comment on CMakeLists file). Only keep active tags and meaningful comments; remove all template and instructional comments to keep the file lean and easy to edit.

islamelnabarawy commented:
Dependency note: This pull request introduces a dependency on [the "joy" ROS package](http://wiki.ros.org/joy). It can be installed in Ubuntu using "apt-get install ros-indigo-joy".

islamelnabarawy commented on ros_ws/src/control/include/control/remote_control.h:
```diff
@@ -0,0 +1,87 @@
+   *
+   * @param msg The message which is received from the joy publisher
+   */
+  void joyCallback(const sensor_msgs::Joy::ConstPtr& msg);
```
This callback is an implementation detail and should be private.

islamelnabarawy commented on ros_ws/src/control/src/remote_control.cpp:
```diff
@@ -0,0 +1,49 @@
+
+  joy_sub = nh.subscribe<sensor_msgs::Joy>("joy", 1, &RemoteControl::joyCallback,
+                                           this);
+  twist_pub = nh.advertise<geometry_msgs::Twist>("teleop_twist", 1);
```
"teleop_twist" should be a constant. Suggested name is TELEOP_TOPIC

islamelnabarawy commented on ros_ws/src/control/src/remote_control_node.cpp:
```diff
@@ -0,0 +1,23 @@
+
+int main(int argc, char** argv)
+{
+  ros::init(argc, argv, "teleop_control");
```
"teleop_control" should be a constant. Suggested name is TELEOP_NODE

islamelnabarawy commented:
Please address all the above comments, and change indentation to use 4 spaces instead of 2, then re-assign pull request to me.

Pull request to be approved after all comments are addressed.

matthewia94 committed:
Remove comment lines from CMakeLists.txt

The auto generated CMakeLists.txt outlines a variety of use cases in the
form of comments which make the file very hard to read when they are not
all used.  The extra unused comments have been removed to make the file
easier to manipulate and read.

matthewia94 committed:
Clean up package.xml

The comments in package.xml made it incredibly hard to read and the
information was also not usable.  The file has been updated with correct
information and the comments have been removed to make it more readable.

matthewia94 commented:
Did some tests and it turns out joy isn't necessary since the actual message is contained in the sensor_msgs package.  I removed it from the package.xml and CMakeLists.txt since it isn't necessarily a dependency.

matthewia94 committed:
Remove joy package dependency

The package only uses the Joy message which is contained in sensor_msgs
and so the joy package is not a dependency.  It has been removed from
both the package.xml and CMakeLists.txt files.

islamelnabarawy commented:
Don't we need to load the actual joy node in the launch file to get the
joystick input messages though?

matthewia94 committed:
Change tabbing to 4 spaces

I have changed all tabbing to be 4 spaces to ensure consistency
throughout the project.

matthewia94 commented:
It doesn't necessarily have to be at the moment.  We could potentially (probably not) write our own controller through say Android and then joy would never be necessary.  I think leave it out for now since it isn't used at the moment.

matthewia94 committed:
Create constants

All of the magic numbers and strings have been replaced with constants
to make the code easier to read and maintain.

islamelnabarawy commented:
I agree that they shouldn't be a build dependency. The implicit runtime
dependency will depend on the controller itself, and the package name has
already been noted here so we should be good.

islamelnabarawy commented on ros_ws/src/control/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+
+################################################
+## Declare ROS dynamic reconfigure parameters ##
+################################################
```
Do we want to keep empty section headers?

islamelnabarawy commented on ros_ws/src/control/include/control/remote_control.h:
```diff
@@ -0,0 +1,99 @@
+/**
+ * @brief The topic that twist messages will be published on
+ */
+extern const std::string TELEOP_TOPIC;
```
I am generally apprehensive of declaring constants in the global namespace. This could potentially cause name collisions in the future. How I would rather handle this is by either scoping these constants into a "constants" namespace (or similar), or create a class for constants and collecting these in there. I'm leaning more towards the second option, but first I'm curious to know what you think.

islamelnabarawy commented on ros_ws/src/control/include/control/remote_control.h:
```diff
@@ -0,0 +1,99 @@
+#include <geometry_msgs/Twist.h>
+
+#define Y_AXIS_L_STICK 1
+#define X_AXIS_R_STICK 2
```
In more modern use of C++, using #define for constants is generally discouraged. Compiler preprocessor directives are better suited to compilation-related tasks, such as switching out code across different build targets.

I would rather we use constants here as well, unless you had a specific reason for defining those in this way.

islamelnabarawy commented on ros_ws/src/control/src/remote_control_node.cpp:
```diff
@@ -0,0 +1,24 @@
+
+int main(int argc, char** argv)
+{
+    ros::init(argc, argv, G_TELEOP_TOPIC);
```
Is the third argument to ros::init the name of the topic or the node?

matthewia94 replied:
Yes

islamelnabarawy commented on ros_ws/src/control/src/remote_control_node.cpp:
```diff
@@ -0,0 +1,24 @@
+    {
+        con.update();
+        ros::spinOnce();
+    }
```
The variable loop_rate seems to be initialized but never used. Do we need to call loop_rate.sleep() at the end of each iteration of the while loop?

Additionally, the frequency (30) probably needs to be a constant for now, then later it needs to be a parameter (that's a separate task).

matthewia94 committed:
Change directive to constant

Adjusted the controller axis indices to be constants rather than
preprocessor directives.

matthewia94 committed:
Remove empty comment headers

There were still some comment headers that were not used and they have
been removed.

matthewia94 committed:
Add sleep to while loop

Added a loop_rate.sleep() to the while loop to ensure that it runs at
the specified rate.

matthewia94 committed:
Add constants.h

Added a constants header which will contain constants which are common
between all packages in the IGVC project.  This includes the loop_rate,
the topic names, the node names and joystick info currently.

matthewia94 commented:
I think that worked well. It seems like it will be hard for you to do all
the reviews by yourself though. Hopefully it will get faster once we have a
good base and we are merging functions rather than packages.  Any thoughts
on the work flow as it stands?

On Sat, Sep 12, 2015, 9:02 PM Islam Elnabarawy notifications@github.com
wrote:

> Merged #7 https://github.com/MST-Robotics/IGVC/pull/7.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/MST-Robotics/IGVC/pull/7#event-407964161.
